File: packages/typescript/src/crs/toWGS84.ts
import { utmData } from "./data";
import type { LatLng } from "../types";

const DEG = Math.PI / 180;

export function toWGS84(easting: number, northing: number, zone: number): LatLng {
  const { a, f, k0, falseEasting, falseNorthingSouth } = utmData.ellipsoid;
  const lambda0 = (zone * 6 - 183) * DEG;

  const e2 = f * (2 - f);
  const ep2 = e2 / (1 - e2);
  const e1 = (1 - Math.sqrt(1 - e2)) / (1 + Math.sqrt(1 - e2));

  const x = easting - falseEasting;
  const y = northing - falseNorthingSouth;

  const M = y / k0;
  const mu =
    M /
    (a * (1 - e2 / 4 - (3 * e2 * e2) / 64 - (5 * e2 * e2 * e2) / 256));

  const phi1 =
    mu +
    ((3 * e1) / 2 - (27 * e1 * e1 * e1) / 32) * Math.sin(2 * mu) +
    ((21 * e1 * e1) / 16 - (55 * e1 * e1 * e1 * e1) / 32) * Math.sin(4 * mu) +
    ((151 * e1 * e1 * e1) / 96) * Math.sin(6 * mu) +
    ((1097 * e1 * e1 * e1 * e1) / 512) * Math.sin(8 * mu);

  const sinPhi1 = Math.sin(phi1);
  const cosPhi1 = Math.cos(phi1);
  const tanPhi1 = Math.tan(phi1);

  const N1 = a / Math.sqrt(1 - e2 * sinPhi1 * sinPhi1);
  const T1 = tanPhi1 * tanPhi1;
  const C1 = ep2 * cosPhi1 * cosPhi1;
  const R1 = (a * (1 - e2)) / Math.pow(1 - e2 * sinPhi1 * sinPhi1, 1.5);
  const D = x / (N1 * k0);

  const D2 = D * D;
  const D3 = D2 * D;
  const D4 = D3 * D;
  const D5 = D4 * D;
  const D6 = D5 * D;

  const phi =
    phi1 -
    ((N1 * tanPhi1) / R1) *
      (D2 / 2 -
        ((5 + 3 * T1 + 10 * C1 - 4 * C1 * C1 - 9 * ep2) * D4) / 24 +
        ((61 + 90 * T1 + 298 * C1 + 45 * T1 * T1 - 252 * ep2 - 3 * C1 * C1) *
          D6) /
          720);

  const lambda =
    lambda0 +
    (D -
      ((1 + 2 * T1 + C1) * D3) / 6 +
      ((5 - 2 * C1 + 28 * T1 - 3 * C1 * C1 + 8 * ep2 + 24 * T1 * T1) * D5) /
        120) /
      cosPhi1;

  return {
    lat: phi / DEG,
    lng: lambda / DEG,
  };
}

File: packages/dart/lib/src/crs/to_wgs84.dart
import 'dart:math' as math;

import 'lat_lng.dart';
import 'utm_config.dart';

LatLng toWGS84(double easting, double northing, int zone) {
  final zoneConfig = UtmConfig.zoneFor(zone);

  final a = UtmConfig.a;
  final f = UtmConfig.f;
  final k0 = UtmConfig.k0;
  final e2 = f * (2 - f);
  final ep2 = e2 / (1 - e2);

  final x = easting - UtmConfig.falseEasting;
  final y = northing - UtmConfig.falseNorthingSouth;

  final m = y / k0;
  final mu = m /
      (a * (1 - e2 / 4 - 3 * e2 * e2 / 64 - 5 * e2 * e2 * e2 / 256));

  final e1 = (1 - math.sqrt(1 - e2)) / (1 + math.sqrt(1 - e2));

  final phi1 = mu +
      (3 * e1 / 2 - 27 * e1 * e1 * e1 / 32) * math.sin(2 * mu) +
      (21 * e1 * e1 / 16 - 55 * e1 * e1 * e1 * e1 / 32) * math.sin(4 * mu) +
      (151 * e1 * e1 * e1 / 96) * math.sin(6 * mu) +
      (1097 * e1 * e1 * e1 * e1 / 512) * math.sin(8 * mu);

  final sinPhi1 = math.sin(phi1);
  final cosPhi1 = math.cos(phi1);
  final tanPhi1 = math.tan(phi1);

  final n1 = a / math.sqrt(1 - e2 * sinPhi1 * sinPhi1);
  final t1 = tanPhi1 * tanPhi1;
  final c1 = ep2 * cosPhi1 * cosPhi1;
  final r1 = a * (1 - e2) / math.pow(1 - e2 * sinPhi1 * sinPhi1, 1.5);
  final d = x / (n1 * k0);

  final phi = phi1 -
      (n1 * tanPhi1 / r1) *
          (d * d / 2 -
              (5 + 3 * t1 + 10 * c1 - 4 * c1 * c1 - 9 * ep2) *
                  math.pow(d, 4) /
                  24 +
              (61 + 90 * t1 + 298 * c1 + 45 * t1 * t1 - 252 * ep2 - 3 * c1 * c1) *
                  math.pow(d, 6) /
                  720);

  final lambda = (d -
          (1 + 2 * t1 + c1) * math.pow(d, 3) / 6 +
          (5 - 2 * c1 + 28 * t1 - 3 * c1 * c1 + 8 * ep2 + 24 * t1 * t1) *
              math.pow(d, 5) /
              120) /
      cosPhi1;

  final lat = phi * 180 / math.pi;
  final lng = zoneConfig.centralMeridian + lambda * 180 / math.pi;

  return LatLng(lat: lat, lng: lng);
}
